### PR TITLE
GOBBLIN-1266: Refactor dataset lineage code to allow Lineage event emission in streaming mode

### PR DESCRIPTION
**Pull request by sv2000**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-1266

### Description
- [x] Here are some details about my PR, including screenshots (if applicable):
Dataset Lineage emission in current Gobblin code is only available for batch mode of execution. For instance, the lineage events are emitted on dataset commit, which is never invoked for long-running tasks. This PR refactors lineage related code so that it is leverageable in streaming mode of execution as well.

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
Existing tests. This PR is purely re-factoring existing code to make it accessible in other classes.

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

**Comment by sv2000** [comment]
@zxcware @autumnust @ZihanLi58 Please review.

File: gobblin-core/src/main/java/org/apache/gobblin/writer/PartitionedDataWriter.java [hunk 1]
```diff
@@ -411,7 +415,7 @@ public void handleMessage(ControlMessage message) {
   /**
    * Get the serialized key to partitions info in {@link #state}
    */
-  private static String getPartitionsKey(int branchId) {
+  public static String getPartitionsKey(int branchId) {
```
Comment (autumnust) [inline gobblin-core/src/main/java/org/apache/gobblin/writer/PartitionedDataWriter.java:418]: why changing to public ?

File: gobblin-metrics-libs/gobblin-metrics-base/src/main/java/org/apache/gobblin/metrics/event/lineage/LineageInfo.java [hunk 2]
```diff
@@ -303,4 +306,50 @@ private static String getKey(Object... objects) {
...
+  private static String getDestinationKey(int branchId) {
+    return getKey(BRANCH, branchId, LineageEventBuilder.DESTINATION);
+  }
+  /**
+   * Remove the destination property from the state object. Used in streaming mode, where we want to selectively purge
+   * lineage information from the state.
+   * @param state
+   * @param branchId
+   */
+  public static void removeDestinationProp(State state, int branchId) {
```
Comment (autumnust) [inline gobblin-metrics-libs/gobblin-metrics-base/src/main/java/org/apache/gobblin/metrics/event/lineage/LineageInfo.java:319]: Where is this being used?

File: gobblin-metrics-libs/gobblin-metrics-base/src/main/java/org/apache/gobblin/metrics/event/lineage/LineageInfo.java [hunk 3]
```diff
@@ -303,4 +306,50 @@ private static String getKey(Object... objects) {
...
+      state.removeProp(destinationKey);
+    }
+  }
+
+  /**
+   * Group states by lineage event name (i.e the dataset name). Used for de-duping LineageEvents for a given dataset.
+   * @param states
+   * @return a map of {@link WorkUnitState}s keyed by dataset name.
+   */
+  public static Map<String, Collection<WorkUnitState>> aggregateByLineageEvent(Collection<? extends WorkUnitState> states) {
```
Comment (autumnust) [inline gobblin-metrics-libs/gobblin-metrics-base/src/main/java/org/apache/gobblin/metrics/event/lineage/LineageInfo.java:331]: Is the signature change necessary ?

**Review by autumnust — CHANGES_REQUESTED** [review]

File: gobblin-core/src/main/java/org/apache/gobblin/writer/PartitionedDataWriter.java (hunk 1 above)
Comment (sv2000) [reply gobblin-core/src/main/java/org/apache/gobblin/writer/PartitionedDataWriter.java:418]: This is needed in LinkedIn's internal publisher. Hence the change.

File: gobblin-metrics-libs/gobblin-metrics-base/src/main/java/org/apache/gobblin/metrics/event/lineage/LineageInfo.java (hunk 2 above)
Comment (sv2000) [reply gobblin-metrics-libs/gobblin-metrics-base/src/main/java/org/apache/gobblin/metrics/event/lineage/LineageInfo.java:319]: same as above.

File: gobblin-metrics-libs/gobblin-metrics-base/src/main/java/org/apache/gobblin/metrics/event/lineage/LineageInfo.java (hunk 3 above)
Comment (sv2000) [reply gobblin-metrics-libs/gobblin-metrics-base/src/main/java/org/apache/gobblin/metrics/event/lineage/LineageInfo.java:331]: Same comment as above.

**Review by autumnust — APPROVED** [review]
LGTM

**Pull request closed without merging**